### DISCUSSION TITLE
PDF Redaction

### DISCUSSION THREAD

**Discussion by TexFly** [pymupdf/PyMuPDF#2383] [Looking for help] [answered]
Hello,
Is there any way I can check with your library if a PDF has been redacted?
I just need a y/n verification.
Thanks!

**Comment by JorjMcKie** [comment]
All you can do is detect whether redaction annotations **_exists_** ... while they do exist.
No difference to other annotation types here.

If a redaction is being applied, it will be deleted - without a trace.
Along with its removal, other things may have been removed too: text, images or hyperlinks - or not, depending on what had been covered by the redaction's rectangle.

Assuming common sense saving options have been used, there is nothing left in the PDF that betrays what has happened.

> Common sense or reasonable saving options are ones that remove unused and duplicate objects, i.e. `garbage=2` at least.

The existence of previous PDF versions is only possible if those versions are kept by intention, i.e. incremental saves.

There also is the PDF's `/ID` field in the trailer. This contains two hash values, one resembling the original version, the second one may indicate that some change has taken place since creation. But this is a highly unreliable indicator, and may not be used at all.

So the answer to your question is "No", because it is impossible. And that does not depend on a tool.

To convince you: here is an example of a PDF `x.pdf` from which text "word2"  has been redacted away giving `y.pdf`. Both are ASCII text only. Nothing in `y.pdf` betrays that a redaction has taken place.
[redact.zip](https://github.com/pymupdf/PyMuPDF/files/11394529/redact.zip)

**Reply by TexFly** [reply]
Thank you very much. I get it. 

How do I detect whether redaction annotations exists ... while they do exist?

**Reply by JorjMcKie** [reply]
You need to iterate over the pages and within each page look for redaction annotations. Like that
```python
for page in doc:
    for annot in page.annots(types=[fitz.PDF_ANNOT_REDACT]):
        print(f"redact annotation on page {page.number} at {annot.rect}")
```

**Reply by TexFly** [reply]
Hey thank you, I appreciate your help. I will write something this morning and test it!

**Reply by TexFly** [reply]
I tried but unfortunately, I do not get any results (meaning the fitz.PDF_ANNOT_REDACT does not catch anything) . I know the file I'm scanning has a redaction because I put it in there.

What else can I try? Should I try with all the Annotation Types?

Thanks.

**Reply by JorjMcKie** [reply]
yeah, look at the page and load this annotation to see what it actually is.

**Reply by TexFly** [reply]
I have looped through all the Annotation Types, and tried multiple files where I actually see a redaction but nothing gets caught. What else can I try?

**Reply by JorjMcKie** [reply, marked as answer]
Sorry for the delayed reaction.
If you loop through all annotations of a page and do not encounter a redaction, then there is no redaction annotation - no matter what you are seeing on the page.
